src/config/schema.ts

```diff
@@ -0,0 +1,337 @@
+import { PackageMode } from '@opengovsg/formsg-sdk/dist/types'
+import awsInfo from 'aws-info'
+import convict, { Schema } from 'convict'
+import { email, url } from 'convict-format-with-validator'
+import { isNil } from 'lodash'
+import mongodbUri from 'mongodb-uri'
+import validator from 'validator'
+
+import {
+  Environment,
+  IBucketUrlSchema,
+  ICompulsoryVarsSchema,
+  IOptionalVarsSchema,
+  IProdOnlyVarsSchema,
+} from '../types'
+
+import defaults from './defaults'
+
+convict.addFormat(url)
+convict.addFormat(email)
+convict.addFormat({
+  name: 'string[]',
+  validate: (val: string[]) => {
+    if (!Array.isArray(val)) {
+      throw new Error('must be of type Array')
+    }
+    let isAllStrings = val.every((i) => typeof i === 'string')
+    if (!isAllStrings) {
+      throw new Error('Elements must be of type string')
+    }
+  },
+  coerce: (val: string): string[] => {
+    return val.split(',')
+  },
+})
+
+/**
+ * Verifies that S3 bucket url is a valid url with or without trailing slash
+ */
+const validateBucketUrl = (
+  val: string,
+  { isDev, hasTrailingSlash }: { isDev: boolean; hasTrailingSlash: boolean },
+) => {
+  if (!validator.isURL(val, { require_tld: !isDev })) {
+    throw new Error('must be a url')
+  }
+  if (hasTrailingSlash) {
+    if (!/[/]$/.test(val)) {
+      throw new Error('must end with a slash')
+    }
+  } else {
+    if (/[/]$/.test(val)) {
+      throw new Error('must not end with a slash')
+    }
+  }
+}
+
+export const compulsoryVarsSchema: Schema<ICompulsoryVarsSchema> = {
+  awsConfig: {
+    imageS3Bucket: {
+      doc: 'S3 Bucket to upload images to',
+      format: String,
+      default: null,
+      env: 'IMAGE_S3_BUCKET',
+    },
+    logoS3Bucket: {
+      doc: 'S3 Bucket to upload logos to',
+      format: String,
+      default: null,
+      env: 'LOGO_S3_BUCKET',
+    },
+    attachmentS3Bucket: {
+      doc: 'S3 Bucket to upload encrypted attachments to',
+      format: String,
+      default: null,
+      env: 'ATTACHMENT_S3_BUCKET',
+    },
+  },
+  core: {
+    sessionSecret: {
+      doc: 'Session Secret',
+      format: String,
+      default: null,
+      env: 'SESSION_SECRET',
+    },
+  },
+}
+
+export const optionalVarsSchema: Schema<IOptionalVarsSchema> = {
+  appConfig: {
+    title: {
+      doc: 'Application name in window title',
+      format: String,
+      default: defaults.app.name,
```

Comment:
I'm wondering whether it might be better to have the defaults in here instead of in a separate file. Our trailing slash error might have been prevented if we had seen all the strings in one place.

Reply:
Fair point. @karrui thoughts? I'm for putting the defaults in the schema file, since it also allows DRY and the `doc` makes it clear what the default is

Reply:
Go for it, since we now have a schema instead of it being scattered everywhere! Some places in the codebase still imports from `default` though, but maybe those can be moved to `constants` instead, heh

Reply:
done